File: blind-75/search_in_rotated_sorted_arr.java
// leetcode 33

import java.util.*;

class Solution {

	/*
	if nums[beg] < nums[mid] for example [4,5,6,7,0,1,2,3] or [0,1,2,3,4,5,6,7,8]
	now check if nums[beg] <= target < nums[mid] then consider first half else second half
	if nums[beg] > nums[mid] for example [7,0,1,2,3,4,5,6] or [6,7,0,1,2,3,4,5]
	now check if target < nums[mid] means <0,1...num[mid]> range or target >= nums[beg] means <...0,1,2> this range
	then first half else second half needs to be considered
	*/
	public int binarySearch(int[] nums, int target, int beg, int end) {
		int mid = (int)((beg + end) / 2);

		System.out.println(Arrays.toString(nums));
		System.out.println("beg : " + beg + " end: " + end + " mid: " + mid);
		System.out.println("nums[beg]: " + nums[beg] + " nums[end]: " + nums[end] + " nums[mid]: " + nums[mid]);

		if (beg == end) {
			if (nums[beg] == target) {
				return beg;
			}
			return -1;
		} else if (nums[mid] == target) {
			return mid;
		} else if (nums[beg] <= nums[mid]) {
			if (target >= nums[beg] && target < nums[mid]) {
				//first half
				end = mid;
			} else {
				// second half
				beg = mid + 1;
			}
		} else { // nums[beg] > nums[mid]
			if (target < nums[mid] || target >= nums[beg]) {
				//first half
				end = mid;
			} else {
				// second half
				beg = mid + 1;
			}
		}

		return binarySearch(nums, target, beg, end);
	}

	public int search(int[] nums, int target) {
		return binarySearch(nums, target, 0, (nums.length - 1));
	}
}


class Main {
	public static void main(String[] args) {
		Solution obj = new Solution();

		int[] nums = {4,5,6,7,0,1,2};
		int target = 0;
		System.out.println(obj.search(nums, target)); // should print 4

		nums = new int[] {4,5,6,7,0,1,2};
		target = 3;
		System.out.println(obj.search(nums, target)); // should print -1

		nums = new int[] {1};
		target = 0;
		System.out.println(obj.search(nums, target)); // should print -1

		nums = new int[] {7,0,1,2,4,5,6};
		target = 1;
		System.out.println(obj.search(nums, target)); // should print 2

		nums = new int[] {1,3};
		target = 1;
		System.out.println(obj.search(nums, target)); // should print 0

		nums = new int[] {1,3};
		target = 3;
		System.out.println(obj.search(nums, target)); // should print 1
	}
}

File: blind-75/search_in_rotated_sorted_arr.py
# leetcode 33

from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low, high = 0, len(nums) - 1

        while low <= high:
        	mid = (low + high) // 2
 
        	if target == nums[mid]:
        		return mid
        	elif nums[low] <= nums[mid] <= nums[high]:
        		# [1,2,3,4,5]
        		if low == high:
        			if nums[low] == target:
        				return low
        			return -1
        		elif target < nums[mid]:
        			high = mid - 1
        		else:
        			low = mid + 1
        	elif nums[mid] <= nums[low] and nums[mid] < nums[high]:
        		# [4,5,1,2,3] or [5,1,2,3,4]
        		if target > nums[mid] and target <= nums[high]:
        			low = mid + 1
        		else:
        			high = mid - 1
        	elif nums[mid] >= nums[low] and nums[mid] > nums[high]:
        		# [2,3,4,5,1] or [3,4,5,1,2]
        		if target < nums[mid] and target >= nums[low]:
        			high = mid - 1
        		else:
        			low = mid + 1

        return -1


# obj = Solution()

# print(obj.search([4,5,6,7,0,1,2], 0))    #  4
# print(obj.search([4,5,6,7,0,1,2], 3))    # -1
# print(obj.search([1], 0))                # -1
# print(obj.search([5,1,3], 5))            #  0
# print(obj.search([5,1,3], 2))  		     # -1
# print(obj.search([5,1,3], 3))  		     #  2

"""
	Approach : Modified Binary Search
	This problem can be generalized based on the following input array
	[1,2,3,4,5]
	- We simply return mid if target == nums[mid]
	- If the array is not rotated at all or rotated n times then the array would be same 
	as before such as [1,2,3,4,5]. Now in this case if nums[low] <= nums[mid] <= nums[high]
	Now check if low == high then if nums[low] == nums[high] then return low
	Else we do the normal binary search procedure
	- Else if nums[mid] <= nums[low] && nums[mid] < nums[high], means [4,5,1,2,3] or [5,1,2,3,4],
	then we check if the num falls under second sequence means target > nums[mid] && target <= nums[high]
	then set low = mid + 1 else high = mid - 1
	- Else if nums[mid] >= nums[low] && nums[mid] > nums[high] means [2,3,4,5,1] or [3,4,5,1,2],
	then check if that falls under the first pure sequence target < nums[mid] && target >= nums[low]
	then set high = mid - 1 else low = mid + 1
"""
